### ISSUE TITLE
Upgrading to Xcode CLT 11.0 broke GCC 9

### ISSUE THREAD

**Issue by crisluengo** [Homebrew/homebrew-core#40676]
I just upgraded the Xcode Command Line Tools to 11.0. Now I can no longer compile with GCC 9. I get this error with any C++ file that includes `cstdlib`:

```
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/stdlib.h:61,
                 from /usr/local/Cellar/gcc/9.1.0/include/c++/9.1.0/cstdlib:75,
                 from /Users/cris/newdip/include/diplib/library/dimension_array.h:30,
                 from /Users/cris/newdip/include/diplib.h:26,
                 from /Users/cris/newdip/src/analysis/findshift.cpp:21:
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/Availability.h:257:22: error: missing binary operator before token "("
  257 |     #if __has_builtin(__is_target_arch)
      |                      ^
```

It looks like `cstdlib` includes `stdlib.h`, but GCC finds it among Clang's header files, rather than the ones that come with GCC. This version of `stdlib.h` uses some compiler intrinsic that is unique to Clang.

This is the output of `gcc --version`:
```
Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 11.0.0 (clang-1100.0.20.17)
Target: x86_64-apple-darwin18.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

Homebrew is up to date and I have the latest GCC installed through Homebrew.

I have uninstalled GCC and installed it anew, with no change in behavior.

Related issues: [1](https://github.com/Homebrew/homebrew-core/issues/35181), [2](https://github.com/Homebrew/legacy-homebrew/issues/34461)

**Comment by crisluengo** [comment]
Is there no way to get this issue reopened?

GCC should **never** include **any** directory in front of its own standard library directories. In particular, the search order seems to be:

1. `/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/stdlib.h`
2. `/usr/local/Cellar/gcc/9.1.0/include/c++/9.1.0/stdlib.h`

but this should be the other way around.

**Comment by SMillerDev** [comment]
Pull requests would be appreciated but there simply aren't enough maintainers of homebrew to maintain beta versions of macOS/apple tools. If you install an unsupported version you're on your own.

**Comment by crisluengo** [comment]
@SMillerDev I understand that you guys are stretched thin. However, this was a normal upgrade suggested by my OS, how am I supposed to know it is an unsupported version?

I'm trying to figure out what is going on. It seems that GCC depends on the system's C library, so this is intended behaviour. It is a change to the system's C library that is not compatible with GCC.

**Comment by crisluengo** [comment]
This is a bug in the C library that comes with the updated Xcode. To fix it, edit the file `/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/Availability.h` (must be done as root). Near the point where the error is generated (line number 257), add the following two lines:

<pre>
#elif defined(__MAC_OS_X_VERSION_MIN_REQUIRED)
  #if defined(__has_builtin) <b>&lt;=== ADD THIS LINE</b>
    #if __has_builtin(__is_target_arch)
     #if __has_builtin(__is_target_vendor)
      #if __has_builtin(__is_target_os)
       #if __has_builtin(__is_target_environment)
        #if __has_builtin(__is_target_variant_os)
         #if __has_builtin(__is_target_variant_environment)
          #if (__is_target_arch(x86_64) && __is_target_vendor(apple) && ((__is_target_os(ios) && __is_target_environment(macabi)) || (__is_target_variant_os(ios) && __is_target_variant_environment(macabi))))
            #define __OSX_AVAILABLE_STARTING(_osx, _ios) __AVAILABILITY_INTERNAL##_osx __AVAILABILITY_INTERNAL##_ios
            #define __OSX_AVAILABLE_BUT_DEPRECATED(_osxIntro, _osxDep, _iosIntro, _iosDep) \
                                                            __AVAILABILITY_INTERNAL##_osxIntro##_DEP##_osxDep __AVAILABILITY_INTERNAL##_iosIntro##_DEP##_iosDep
            #define __OSX_AVAILABLE_BUT_DEPRECATED_MSG(_osxIntro, _osxDep, _iosIntro, _iosDep, _msg) \
                                                            __AVAILABILITY_INTERNAL##_osxIntro##_DEP##_osxDep##_MSG(_msg) __AVAILABILITY_INTERNAL##_iosIntro##_DEP##_iosDep##_MSG(_msg)
          #endif /* # if __is_target_arch... */
         #endif /* #if __has_builtin(__is_target_variant_environment) */
        #endif /* #if __has_builtin(__is_target_variant_os) */
       #endif /* #if __has_builtin(__is_target_environment) */
      #endif /* #if __has_builtin(__is_target_os) */
     #endif /* #if __has_builtin(__is_target_vendor) */
    #endif /* #if __has_builtin(__is_target_arch) */
  #endif  <b>&lt;=== ADD THIS LINE</b>

    #ifndef __OSX_AVAILABLE_STARTING
        #define __OSX_AVAILABLE_STARTING(_osx, _ios) __AVAILABILITY_INTERNAL##_osx
        #define __OSX_AVAILABLE_BUT_DEPRECATED(_osxIntro, _osxDep, _iosIntro, _iosDep) \
                                                        __AVAILABILITY_INTERNAL##_osxIntro##_DEP##_osxDep
        #define __OSX_AVAILABLE_BUT_DEPRECATED_MSG(_osxIntro, _osxDep, _iosIntro, _iosDep, _msg) \
                                                        __AVAILABILITY_INTERNAL##_osxIntro##_DEP##_osxDep##_MSG(_msg)
    #endif /* __OSX_AVAILABLE_STARTING */

#else
</pre>

Needless to say, edit system files at your own risk! :)

**Comment by bermudalocket** [comment]
@crisluengo unfortunately the issue persists for me even after making that change. I'm on macOS 10.15 though and made the above suggested edit in the 10.15 SDK instead of the 10.14 SDK. I've been debugging this since Monday and was hoping this was the solution. Frustrating :(

**Comment by fxcoudert** [comment]
> this was a normal upgrade suggested by my OS, how am I supposed to know it is an unsupported version?

Xcode 11 is beta, so this was not a “normal” update, or you've configured at some point your system to install beta versions (and likely forgotten about it 😄).

> GCC should never include any directory in front of its own standard library directories

This is not the issue.

> It is a change to the system's C library that is not compatible with GCC

Yes. The beta SDK includes in `Availability.h` some code that is not standard C compliant, because it uses the clang-specific `__has_builtin` built-in unconditionally. Please report this to Apple so they can get it fixed.

**Comment by pronvit** [comment]
`sys/cdefs.h` defines `__has_builtin` and others if they're not available. Including `sys/cdefs.h` before `stdlib.h` fixes the problem.

**Comment by fxcoudert** [comment]
> Including sys/cdefs.h before stdlib.h fixes the problem

There are many such workarounds. The only real fix is to report the problem to Apple, so they can fix it. (Duplicate reports actually help them determine what is a pressing issue, too.)

**Comment by QYQSDTC** [comment]
> This is a bug in the C library that comes with the updated Xcode. To fix it, edit the file `/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/Availability.h` (must be done as root). Near the point where the error is generated (line number 257), add the following two lines:
> 
> #elif defined(__MAC_OS_X_VERSION_MIN_REQUIRED)
>   #if defined(__has_builtin)   // <=== ADD THIS LINE
>     #if __has_builtin(__is_target_arch)
>      #if __has_builtin(__is_target_vendor)
>       #if __has_builtin(__is_target_os)
>        #if __has_builtin(__is_target_environment)
>         #if __has_builtin(__is_target_variant_os)
>          #if __has_builtin(__is_target_variant_environment)
>           #if (__is_target_arch(x86_64) && __is_target_vendor(apple) && ((__is_target_os(ios) && __is_target_environment(macabi)) || (__is_target_variant_os(ios) && __is_target_variant_environment(macabi))))
>             #define __OSX_AVAILABLE_STARTING(_osx, _ios) __AVAILABILITY_INTERNAL##_osx __AVAILABILITY_INTERNAL##_ios
>             #define __OSX_AVAILABLE_BUT_DEPRECATED(_osxIntro, _osxDep, _iosIntro, _iosDep) \
>                                                             __AVAILABILITY_INTERNAL##_osxIntro##_DEP##_osxDep __AVAILABILITY_INTERNAL##_iosIntro##_DEP##_iosDep
>             #define __OSX_AVAILABLE_BUT_DEPRECATED_MSG(_osxIntro, _osxDep, _iosIntro, _iosDep, _msg) \
>                                                             __AVAILABILITY_INTERNAL##_osxIntro##_DEP##_osxDep##_MSG(_msg) __AVAILABILITY_INTERNAL##_iosIntro##_DEP##_iosDep##_MSG(_msg)
>           #endif /* # if __is_target_arch... */
>          #endif /* #if __has_builtin(__is_target_variant_environment) */
>         #endif /* #if __has_builtin(__is_target_variant_os) */
>        #endif /* #if __has_builtin(__is_target_environment) */
>       #endif /* #if __has_builtin(__is_target_os) */
>      #endif /* #if __has_builtin(__is_target_vendor) */
>     #endif /* #if __has_builtin(__is_target_arch) */
>   #endif   // <=== ADD THIS LINE
> 
>     #ifndef __OSX_AVAILABLE_STARTING
>         #define __OSX_AVAILABLE_STARTING(_osx, _ios) __AVAILABILITY_INTERNAL##_osx
>         #define __OSX_AVAILABLE_BUT_DEPRECATED(_osxIntro, _osxDep, _iosIntro, _iosDep) \
>                                                         __AVAILABILITY_INTERNAL##_osxIntro##_DEP##_osxDep
>         #define __OSX_AVAILABLE_BUT_DEPRECATED_MSG(_osxIntro, _osxDep, _iosIntro, _iosDep, _msg) \
>                                                         __AVAILABILITY_INTERNAL##_osxIntro##_DEP##_osxDep##_MSG(_msg)
>     #endif /* __OSX_AVAILABLE_STARTING */
> 
> #else
> Needless to say, edit system files at your own risk! :)

Bravo dude!!! I updated yesterday and has this problem also.

**Comment by QYQSDTC** [comment]
@crisluengo It fixes my problem, I just updated yesterday. Thanks bro.

**Comment by dgboy2000** [comment]
This same issue happens for me in Xcode 10.2.1. @crisluengo's solution of ordering import paths properly makes the most sense to me. I was able to ditch homebrew's gcc and use clang but wanted to "+1" this post.

**Comment by ggouaillardet** [comment]
TL;DR I faced the same issue and `configure CPPFLAGS='-D__has_builtin(x)=0' ...` was my workaround.

On my up-to-date (OS X and brew) Mac (with zero beta software) I tried to build [Open MPI](https://open-mpi.org) and `configure` failed with the following error message
```
ptrdiff_t type is not available, this is required by C99 standard.  Cannot continue
```

after some debugging, I found that this `autoconf` test failed
```
checking for ANSI C header files... no
```

and this is caused by the issue discussed here.


Though it is technically correct to blame Apple for breaking `stdlib.h`, or even Open MPI for expecting `autoconf` will `#include <stddef.h>` under the hood, I am afraid this issue has a much larger impact that some might have expected.
I do not know the policy (if any) for dealing with such cases, but IMHO, a workaround should be included in the brew package in order to be (even) more user friendly.

my 0.02US$

**Comment by jeffhammond** [comment]
I see this with Xcode Version 10.2.1 (10E1001) but find that I can resolve it by switching
```c
#include <stdlib.h>
#include <stdio.h>
```
to
```c
#include <stdio.h>
#include <stdlib.h>
```
[MCVE](https://github.com/jeffhammond/HPCInfo/tree/master/bugs/macosx/theworst)

**Comment by ggouaillardet** [comment]
FWIW, based on @jeffhammond workaround, I made the patch below for autoconf so `stdio.h` is always included before `stdlib.h` is used.

Though the patch itself should not fix anything, it is harmless and good enough (for now) to work around the broken headers.

@MikeMcQuaid FYI

in order to apply it, simply add the following lines in the `autoconf` formula
```
  patch do
    url "https://gist.githubusercontent.com/ggouaillardet/11e15a9349dc2100d98d59a5ccc3c74e/raw/34e16d46fd1f1f5cc060d15afbb3c94a385ebdcb/autoconf.diff"
    sha256 "7c415d8994fa668dc16803950c0b3e92ad8f400fbf5a7adc0f6ea59fe29a3640"
  end
```

```patch
diff -ruN orig/autoconf-2.69/lib/autoconf/c.m4 autoconf-2.69/lib/autoconf/c.m4
--- orig/autoconf-2.69/lib/autoconf/c.m4	2012-01-21 22:46:39.000000000 +0900
+++ autoconf-2.69/lib/autoconf/c.m4	2019-07-22 17:01:36.000000000 +0900
@@ -1204,11 +1204,11 @@
 # variable length arrays.
 AC_DEFUN([_AC_PROG_CC_C99],
 [_AC_C_STD_TRY([c99],
-[[#include <stdarg.h>
+[[#include <stdio.h>
+#include <stdarg.h>
 #include <stdbool.h>
 #include <stdlib.h>
 #include <wchar.h>
-#include <stdio.h>
 
 // Check varargs macros.  These examples are taken from C99 6.10.3.5.
 #define debug(...) fprintf (stderr, __VA_ARGS__)
@@ -1849,8 +1849,8 @@
     ac_cv_c_flexmember,
     [AC_COMPILE_IFELSE(
        [AC_LANG_PROGRAM(
-	  [[#include <stdlib.h>
-	    #include <stdio.h>
+	  [[#include <stdio.h>
+	    #include <stdlib.h>
 	    #include <stddef.h>
 	    struct s { int n; double d[]; };]],
 	  [[int m = getchar ();
diff -ruN orig/autoconf-2.69/lib/autoconf/functions.m4 autoconf-2.69/lib/autoconf/functions.m4
--- orig/autoconf-2.69/lib/autoconf/functions.m4	2012-01-21 22:46:39.000000000 +0900
+++ autoconf-2.69/lib/autoconf/functions.m4	2019-07-22 17:03:44.000000000 +0900
@@ -890,7 +890,8 @@
 AC_CACHE_CHECK([for GNU libc compatible malloc], ac_cv_func_malloc_0_nonnull,
 [AC_RUN_IFELSE(
 [AC_LANG_PROGRAM(
-[[#if defined STDC_HEADERS || defined HAVE_STDLIB_H
+[[#include <stdio.h>
+#if defined STDC_HEADERS || defined HAVE_STDLIB_H
 # include <stdlib.h>
 #else
 char *malloc ();
@@ -1003,6 +1004,7 @@
 #endif
 
 #include <limits.h>
+#include <stdio.h>
 #include <stdlib.h>
 
 #ifdef HAVE_UNISTD_H
@@ -1395,7 +1397,8 @@
 AC_CACHE_CHECK([for GNU libc compatible realloc], ac_cv_func_realloc_0_nonnull,
 [AC_RUN_IFELSE(
 [AC_LANG_PROGRAM(
-[[#if defined STDC_HEADERS || defined HAVE_STDLIB_H
+[[#include <stdio.h>
+#if defined STDC_HEADERS || defined HAVE_STDLIB_H
 # include <stdlib.h>
 #else
 char *realloc ();
@@ -1595,6 +1598,7 @@
 	  [[/* On HP-UX before 11.23, strtold returns a struct instead of
 		long double.  Reject implementations like that, by requiring
 		compatibility with the C99 prototype.  */
+#            include <stdio.h>
 #	     include <stdlib.h>
 	     static long double (*p) (char const *, char **) = strtold;
 	     static long double
diff -ruN orig/autoconf-2.69/lib/autoconf/headers.m4 autoconf-2.69/lib/autoconf/headers.m4
--- orig/autoconf-2.69/lib/autoconf/headers.m4	2012-04-25 11:37:26.000000000 +0900
+++ autoconf-2.69/lib/autoconf/headers.m4	2019-07-22 17:02:16.000000000 +0900
@@ -676,7 +676,8 @@
 # --------------
 AC_DEFUN([AC_HEADER_STDC],
 [AC_CACHE_CHECK(for ANSI C header files, ac_cv_header_stdc,
-[AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[#include <stdlib.h>
+[AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[#include <stdio.h>
+#include <stdlib.h>
 #include <stdarg.h>
 #include <string.h>
 #include <float.h>
@@ -697,7 +698,8 @@
 if test $ac_cv_header_stdc = yes; then
   # /bin/cc in Irix-4.0.5 gets non-ANSI ctype macros unless using -ansi.
   AC_RUN_IFELSE([AC_LANG_SOURCE(
-[[#include <ctype.h>
+[[#include <stdio.h>
+#include <ctype.h>
 #include <stdlib.h>
 #if ((' ' & 0x0FF) == 0x020)
 # define ISLOWER(c) ('a' <= (c) && (c) <= 'z')
```

**Comment by SMillerDev** [comment]
Thanks for the work but I'm just gonna repeat what fx said cause it's essential in this:

> There are many such workarounds. The only real fix is to report the problem to Apple, so they can fix it. (Duplicate reports actually help them determine what is a pressing issue, too.)

**Comment by jeffhammond** [comment]
One can dismiss essentially all good engineering solutions by identifying better ones.  
https://en.wikipedia.org/wiki/Perfect_is_the_enemy_of_good

**Comment by MikeMcQuaid** [comment]
@jeffhammond Your meta-commentary is not welcome or necessary, thanks.

**Comment by timkphd** [comment]
I'm using gcc and fortran from http://hpc.sourceforge.net, Xcode 11.3, macOS 10.14.6.  Company policy prevents me from upgrading macOS.  

The suggestion to modify
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/Availability.h

#if defined(__has_builtin)   // <=== ADD THIS LINE
...
 #endif   // <=== ADD THIS LINE

did not work for me.

This might break something along the way but so far I have been able to build several things including OpenMPI 4.xx...

I added 

#include <stdio.h>

as the first "real" line of  

/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/Availability.h
